packages/react/src/host/RemoteComponent.tsx

```diff
@@ -41,21 +41,23 @@ export const RemoteComponent = memo(
     return (
       <Implementation {...props}>
         {[...children].map((child) => {
-          if ('children' in child) {
-            return (
-              <RemoteComponent
-                key={child.id}
-                receiver={receiver}
-                component={child}
-                controller={controller}
-                __type__={(controller.get(child.type) as any)?.__type__}
-              />
-            );
-          } else {
-            return (
-              <RemoteText key={child.id} text={child} receiver={receiver} />
-            );
+          let element: ReactElement | null;
+          switch (child.kind) {
+            case KIND_COMPONENT:
+              element = renderComponent({
+                component: child,
+                receiver,
+                controller,
+              });
+              break;
+            case KIND_TEXT:
+              element = renderText({text: child, receiver});
+              break;
+            default:
+              element = null;
+              break;
           }
+          return element ? cloneElement(element, {key: child.id}) : null;
```

Comment:
Do we have to clone the element, or could the `renderText`/ `renderComponent` call be expected to do that? This component is rendered for every component in the tree so keeping it as lightweight as possible is definitely desirable

Reply:
Can't `renderComponent` add the key?